dbms/src/Flash/Mpp/ExchangeReceiver.cpp

```diff
@@ -967,21 +1042,38 @@ void ExchangeReceiverBase<RPCContext>::connectionLocalDone()
 {
     std::lock_guard lock(mu);
     --live_local_connections;
-    cv.notify_all();
+    if (live_local_connections == 0)
+        cv.notify_all();
 }
 
 template <typename RPCContext>
 void ExchangeReceiverBase<RPCContext>::finishAllMsgChannels()
 {
-    for (auto & msg_channel : msg_channels)
-        msg_channel->finish();
+    if (grpc_recv_queue.empty())
+    {
+        for (auto & msg_channel : msg_channels)
+            msg_channel->finish();
+    }
+    else
+    {
+        for (auto & channel : grpc_recv_queue)
+            channel.finish();
+    }
 }
 
 template <typename RPCContext>
 void ExchangeReceiverBase<RPCContext>::cancelAllMsgChannels()
 {
-    for (auto & msg_channel : msg_channels)
-        msg_channel->cancel();
+    if (grpc_recv_queue.empty())
+    {
+        for (auto & msg_channel : msg_channels)
+            msg_channel->cancel();
+    }
+    else
+    {
+        for (auto & channel : grpc_recv_queue)
+            channel.cancel();
+    }
```

Comment:
Looks like `grpc_recv_queue` won't be empty is it?

Reply:
> Looks like `grpc_recv_queue` won't be empty is it?

Yes, and I will change the logic related to `grpc_recv_queue`.